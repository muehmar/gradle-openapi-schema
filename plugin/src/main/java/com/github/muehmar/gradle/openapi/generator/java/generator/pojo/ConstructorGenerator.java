package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.util.Functions.allExceptFirst;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.model.name.JavaName;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.JavaModifier;
import io.github.muehmar.codegenerator.java.JavaModifiers;
import io.github.muehmar.codegenerator.writer.Writer;
import io.github.muehmar.pojobuilder.annotations.FieldBuilder;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.UnaryOperator;

@PojoBuilder
public class ConstructorGenerator<A, B> implements Generator<A, B> {
  private final BiFunction<A, B, JavaModifiers> createModifiers;
  private final BiFunction<A, B, String> createClassName;
  private final BiFunction<A, B, PList<String>> createArguments;
  private final Generator<A, B> contentGenerator;

  ConstructorGenerator(
      BiFunction<A, B, JavaModifiers> createModifiers,
      BiFunction<A, B, String> createClassName,
      BiFunction<A, B, PList<String>> createArguments,
      Generator<A, B> contentGenerator) {
    this.createModifiers = createModifiers;
    this.createClassName = createClassName;
    this.createArguments = createArguments;
    this.contentGenerator = contentGenerator;
  }

  @Override
  public Writer generate(A data, B settings, Writer writer) {
    return Generator.<A, B>emptyGen()
        .append(
            w -> {
              final JavaModifiers modifiers = createModifiers.apply(data, settings);
              final String className = createClassName.apply(data, settings);
              return w.println("%s%s(", modifiers.asStringTrailingWhitespace(), className);
            })
        .append(argumentGenerator(), 2)
        .append(w -> w.tab(1).println(") {"))
        .append(contentGenerator, 1)
        .append(w -> w.println("}"))
        .generate(data, settings, writer);
  }

  private Generator<A, B> argumentGenerator() {
    return (data, settings, writer) ->
        createArguments
            .apply(data, settings)
            .reverse()
            .zipWithIndex()
            .map(allExceptFirst(arg -> arg + ","))
            .reverse()
            .foldLeft(writer, Writer::println);
  }

  @FieldBuilder(fieldName = "createModifiers")
  static class ModifiersGen {
    private ModifiersGen() {}

    static <A, B> BiFunction<A, B, JavaModifiers> modifiers() {
      return (d, s) -> JavaModifiers.of();
    }

    static <A, B> BiFunction<A, B, JavaModifiers> modifiers(JavaModifier modifier) {
      return (d, s) -> JavaModifiers.of(modifier);
    }

    static <A, B> BiFunction<A, B, JavaModifiers> modifiers(Function<A, JavaModifiers> modifier) {
      return (d, s) -> modifier.apply(d);
    }

    static <A, B> BiFunction<A, B, JavaModifiers> modifiers(JavaModifier m1, JavaModifier m2) {
      return (d, s) -> JavaModifiers.of(m1, m2);
    }
  }

  @FieldBuilder(fieldName = "createClassName")
  static class ClassNameBuilder {
    private ClassNameBuilder() {}

    static <A, B> BiFunction<A, B, String> className(BiFunction<A, B, String> createClassName) {
      return createClassName;
    }

    static <A, B> BiFunction<A, B, String> javaClassName(Function<A, JavaName> createClassName) {
      return (d, s) -> createClassName.apply(d).asString();
    }

    static <A, B> BiFunction<A, B, String> className(Function<A, String> createClassName) {
      return (d, s) -> createClassName.apply(d);
    }

    static <A, B> BiFunction<A, B, String> className(String className) {
      return (d, s) -> className;
    }
  }

  @FieldBuilder(fieldName = "createArguments")
  static class ArgumentsBuilder {
    private ArgumentsBuilder() {}

    static <A, B> BiFunction<A, B, PList<String>> arguments(
        BiFunction<A, B, PList<String>> createArguments) {
      return createArguments;
    }

    static <A, B> BiFunction<A, B, PList<String>> arguments(
        Function<A, PList<String>> createArguments) {
      return (d, s) -> createArguments.apply(d);
    }

    static <A, B> BiFunction<A, B, PList<String>> singleArgument(
        Function<A, String> createArgument) {
      return (d, s) -> PList.single(createArgument.apply(d));
    }

    static <A, B> BiFunction<A, B, PList<String>> singleArgument(String argument) {
      return (d, s) -> PList.single(argument);
    }

    static <A, B> BiFunction<A, B, PList<String>> noArguments() {
      return (d, s) -> PList.empty();
    }
  }

  @FieldBuilder(fieldName = "contentGenerator")
  static class ContentBuilder {
    private ContentBuilder() {}

    static <A, B> Generator<A, B> content(String content) {
      return (data, settings, writer) -> writer.println(content);
    }

    static <A, B> Generator<A, B> content(Generator<A, B> content) {
      return content;
    }

    static <A, B> Generator<A, B> content(UnaryOperator<Writer> content) {
      return (data, settings, writer) -> content.apply(writer);
    }

    static <A, B> Generator<A, B> noContent() {
      return (data, settings, writer) -> writer;
    }
  }
}
