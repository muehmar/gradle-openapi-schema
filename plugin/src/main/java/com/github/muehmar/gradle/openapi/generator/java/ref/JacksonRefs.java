package com.github.muehmar.gradle.openapi.generator.java.ref;

public class JacksonRefs {
  private JacksonRefs() {}

  public static final String JSON_POJO_BUILDER =
      "com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder";
  public static final String JSON_DESERIALIZE =
      "com.fasterxml.jackson.databind.annotation.JsonDeserialize";
  public static final String JSON_VALUE = "com.fasterxml.jackson.annotation.JsonValue";
  public static final String JSON_CREATOR = "com.fasterxml.jackson.annotation.JsonCreator";
  public static final String JSON_ANY_GETTER = "com.fasterxml.jackson.annotation.JsonAnyGetter";
  public static final String JSON_ANY_SETTER = "com.fasterxml.jackson.annotation.JsonAnySetter";
  public static final String JSON_IGNORE = "com.fasterxml.jackson.annotation.JsonIgnore";
  public static final String JSON_INCLUDE = "com.fasterxml.jackson.annotation.JsonInclude";
  public static final String JSON_PROPERTY = "com.fasterxml.jackson.annotation.JsonProperty";
  public static final String JACKSON_XML_ROOT_ELEMENT =
      "com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlRootElement";
  public static final String JACKSON_XML_PROPERTY =
      "com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty";
  public static final String JACKSON_XML_ELEMENT_WRAPPER =
      "com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlElementWrapper";
}
