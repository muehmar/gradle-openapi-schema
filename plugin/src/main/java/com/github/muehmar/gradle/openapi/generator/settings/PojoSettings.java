package com.github.muehmar.gradle.openapi.generator.settings;

import static com.github.muehmar.gradle.openapi.util.Booleans.not;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.model.name.Name;
import com.github.muehmar.gradle.openapi.task.TaskIdentifier;
import com.github.muehmar.gradle.openapi.warnings.Warning;
import com.github.muehmar.gradle.openapi.warnings.WarningsContext;
import io.github.muehmar.pojobuilder.annotations.FieldBuilder;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.io.Serializable;
import java.util.List;
import lombok.Value;
import lombok.With;

@Value
@With
@PojoBuilder
public class PojoSettings implements Serializable {
  JsonSupport jsonSupport;
  String packageName;
  String suffix;
  StagedBuilderSettings stagedBuilder;
  String builderMethodPrefix;
  boolean enableValidation;
  boolean nonStrictOneOfValidation;

  ValidationApi validationApi;
  List<ClassTypeMapping> classTypeMappings;
  List<FormatTypeMapping> formatTypeMappings;
  EnumDescriptionSettings enumDescriptionSettings;
  GetterSuffixes getterSuffixes;
  ValidationMethods validationMethods;
  List<String> excludeSchemas;
  PojoNameMappings pojoNameMappings;

  TaskIdentifier taskIdentifier;

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableValidation() {
    return enableValidation;
  }

  public boolean isNonStrictOneOfValidation() {
    return nonStrictOneOfValidation;
  }

  public List<ClassTypeMapping> getClassTypeMappingsList() {
    return classTypeMappings;
  }

  public List<FormatTypeMapping> getFormatTypeMappingsList() {
    return formatTypeMappings;
  }

  public List<String> getExcludeSchemasList() {
    return excludeSchemas;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public ExcludedSchemas getExcludedSchemas() {
    return ExcludedSchemas.fromExcludedPojoNames(
        PList.fromIter(excludeSchemas).map(Name::ofString));
  }

  public boolean isEnableStagedBuilder() {
    return stagedBuilder.isEnabled();
  }

  public boolean isDisableStagedBuilder() {
    return !isEnableStagedBuilder();
  }

  public StagedBuilderSettings getStagedBuilder() {
    return stagedBuilder;
  }

  @FieldBuilder(fieldName = "classTypeMappings")
  public static List<ClassTypeMapping> classTypeMappings(
      PList<ClassTypeMapping> classTypeMappings) {
    return classTypeMappings.toArrayList();
  }

  @FieldBuilder(fieldName = "formatTypeMappings")
  public static List<FormatTypeMapping> formatTypeMappings(
      PList<FormatTypeMapping> formatTypeMappings) {
    return formatTypeMappings.toArrayList();
  }

  public TypeMappings getTypeMappings() {
    return new TypeMappings(PList.fromIter(classTypeMappings), PList.fromIter(formatTypeMappings));
  }

  public PojoNameMapping pojoNameMapping() {
    return PList.fromIter(pojoNameMappings.getConstantNameMappings())
        .map(
            constantNameMapping ->
                PojoNameMapping.replaceConstant(
                    constantNameMapping.getConstant(), constantNameMapping.getReplacement()))
        .foldLeft(PojoNameMapping.noMapping(), PojoNameMapping::andThen);
  }

  public void validate() {
    classTypeMappings.forEach(
        classTypeMapping -> {
          if (not(classTypeMapping.getTypeConversion().isPresent())) {
            final Warning warning = Warning.missingMappingConversion(classTypeMapping);
            WarningsContext.addWarningForTask(taskIdentifier, warning);
          }
        });

    formatTypeMappings.forEach(
        formatTypeMapping -> {
          if (not(formatTypeMapping.getTypeConversion().isPresent())) {
            final Warning warning = Warning.missingMappingConversion(formatTypeMapping);
            WarningsContext.addWarningForTask(taskIdentifier, warning);
          }
        });
  }
}
