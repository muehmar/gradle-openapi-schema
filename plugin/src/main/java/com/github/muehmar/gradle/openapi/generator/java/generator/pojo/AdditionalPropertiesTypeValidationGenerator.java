package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.generator.shared.ValidationGenerator.assertTrue;
import static com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties.additionalPropertiesName;
import static io.github.muehmar.codegenerator.Generator.constant;

import com.github.muehmar.gradle.openapi.generator.java.generator.shared.SettingsFunctions;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.MethodGenBuilder;

public class AdditionalPropertiesTypeValidationGenerator {
  private AdditionalPropertiesTypeValidationGenerator() {}

  public static Generator<JavaObjectPojo, PojoSettings>
      additionalPropertiesTypeValidationGenerator() {
    return annotatedCorrectTypeMethod()
        .filter(JavaAdditionalProperties::isAllowed)
        .filter(JavaAdditionalProperties::isNotValueAnyType)
        .contraMap(JavaObjectPojo::getAdditionalProperties);
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> annotatedCorrectTypeMethod() {
    return JacksonAnnotationGenerator.<JavaAdditionalProperties>jsonIgnore()
        .append(
            assertTrue(
                props ->
                    String.format(
                        "Not all additional properties are instances of %s",
                        props.getType().getQualifiedClassName().getClassName())))
        .append(correctTypeMethod());
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> correctTypeMethod() {
    return MethodGenBuilder.<JavaAdditionalProperties, PojoSettings>create()
        .modifiers(SettingsFunctions::validationMethodModifiers)
        .noGenericTypes()
        .returnType("boolean")
        .methodName("isAllAdditionalPropertiesHaveCorrectType")
        .noArguments()
        .content(
            constant(
                String.format(
                    "return getAdditionalProperties().size() == %s.size();",
                    additionalPropertiesName())))
        .build();
  }
}
