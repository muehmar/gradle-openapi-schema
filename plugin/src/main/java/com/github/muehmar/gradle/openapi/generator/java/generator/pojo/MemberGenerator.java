package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties.additionalPropertiesName;
import static io.github.muehmar.codegenerator.Generator.constant;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.member.TechnicalPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.ref.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Optional;
import lombok.Value;

public class MemberGenerator {
  private MemberGenerator() {}

  public static Generator<MemberContent, PojoSettings> memberGenerator() {
    return Generator.<MemberContent, PojoSettings>emptyGen()
        .append(jsonValueAnnotation())
        .appendList(singleMember(), MemberContent::getMembers)
        .appendOptional(additionalPropertiesMember(), MemberContent::getAdditionalProperties);
  }

  private static Generator<TechnicalPojoMember, PojoSettings> singleMember() {
    return Generator.<TechnicalPojoMember, PojoSettings>emptyGen()
        .append(
            (field, settings, writer) ->
                writer.println(
                    "private final %s %s;",
                    field.getJavaType().getParameterizedClassName(), field.getName()))
        .append(RefsGenerator.javaTypeRefs(), TechnicalPojoMember::getJavaType);
  }

  private static Generator<JavaAdditionalProperties, PojoSettings> additionalPropertiesMember() {
    return Generator.<JavaAdditionalProperties, PojoSettings>emptyGen()
        .append(constant("private final Map<String, Object> %s;", additionalPropertiesName()))
        .append(RefsGenerator.ref(JavaRefs.JAVA_UTIL_MAP));
  }

  private static Generator<MemberContent, PojoSettings> jsonValueAnnotation() {
    return JacksonAnnotationGenerator.<MemberContent>jsonValue().filter(MemberContent::isArrayPojo);
  }

  @Value
  @PojoBuilder(builderName = "MemberContentBuilder")
  public static class MemberContent {
    boolean isArrayPojo;
    PList<TechnicalPojoMember> members;
    Optional<JavaAdditionalProperties> additionalProperties;
  }
}
