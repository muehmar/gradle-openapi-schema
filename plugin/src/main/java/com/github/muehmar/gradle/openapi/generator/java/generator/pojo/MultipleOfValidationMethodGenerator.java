package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.Filters;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.SettingsFunctions;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.ValidationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.model.constraints.MultipleOf;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.MethodGen;
import io.github.muehmar.codegenerator.java.MethodGenBuilder;
import lombok.Value;

public class MultipleOfValidationMethodGenerator {
  private MultipleOfValidationMethodGenerator() {}

  public static Generator<JavaObjectPojo, PojoSettings> multipleOfValidationMethodGenerator() {
    return Generator.<JavaObjectPojo, PojoSettings>emptyGen()
        .appendList(memberGenerator().appendSingleBlankLine(), MemberAndConstraint::fromPojo)
        .filter(Filters.isValidationEnabled());
  }

  private static Generator<MemberAndConstraint, PojoSettings> memberGenerator() {
    final MethodGen<MemberAndConstraint, PojoSettings> method =
        MethodGenBuilder.<MemberAndConstraint, PojoSettings>create()
            .modifiers(SettingsFunctions::validationMethodModifiers)
            .noGenericTypes()
            .returnType("boolean")
            .methodName(MemberAndConstraint::getIsMultipleOfValidMethodName)
            .noArguments()
            .content(content())
            .build();

    final Generator<MemberAndConstraint, PojoSettings> assertTrueAnnotation =
        ValidationGenerator.assertTrue(
            mc ->
                String.format(
                    "%s is not a multiple of %s",
                    mc.member.getName(),
                    mc.member
                        .getJavaType()
                        .getConstraints()
                        .getMultipleOf()
                        .map(MultipleOf::asString)
                        .orElse("0")));
    return Generator.<MemberAndConstraint, PojoSettings>emptyGen()
        .append(assertTrueAnnotation)
        .append(method)
        .filter(MemberAndConstraint::isIntegerOrNumericType);
  }

  private static Generator<MemberAndConstraint, PojoSettings> content() {
    return Generator.<MemberAndConstraint, PojoSettings>emptyGen()
        .append(integerAndLongContent())
        .append(floatAndDoubleContent());
  }

  private static Generator<MemberAndConstraint, PojoSettings> integerAndLongContent() {
    return Generator.<MemberAndConstraint, PojoSettings>emptyGen()
        .append(
            (mc, s, w) ->
                w.println(
                    "return %s == null || %s %% %sL == 0;",
                    mc.member.getName(), mc.member.getName(), mc.multipleOf.asString()))
        .filter(MemberAndConstraint::isIntegerType);
  }

  private static Generator<MemberAndConstraint, PojoSettings> floatAndDoubleContent() {
    return Generator.<MemberAndConstraint, PojoSettings>emptyGen()
        .append(
            (mc, s, w) ->
                w.println(
                    "return %s == null || BigDecimal.valueOf(%s).divideAndRemainder(new BigDecimal(\"%s\"))[1].signum() == 0;",
                    mc.member.getName(), mc.member.getName(), mc.multipleOf.asString()))
        .append(w -> w.ref(JavaRefs.JAVA_MATH_BIG_DECIMAL))
        .filter(MemberAndConstraint::isNumericType);
  }

  @Value
  private static class MemberAndConstraint {
    JavaPojoMember member;
    MultipleOf multipleOf;

    static PList<MemberAndConstraint> fromPojo(JavaObjectPojo pojo) {
      return pojo.getMembers()
          .flatMapOptional(
              member ->
                  member
                      .getJavaType()
                      .getConstraints()
                      .getMultipleOf()
                      .map(multipleOf -> new MemberAndConstraint(member, multipleOf)));
    }

    public String getIsMultipleOfValidMethodName() {
      return member.getName().startUpperCase().prefix("is").append("MultipleOfValid").asString();
    }

    public boolean isIntegerType() {
      return member.getJavaType().getType().asIntegerType().isPresent();
    }

    public boolean isNumericType() {
      return member.getJavaType().getType().asNumericType().isPresent();
    }

    public boolean isIntegerOrNumericType() {
      return isIntegerType() || isNumericType();
    }
  }
}
