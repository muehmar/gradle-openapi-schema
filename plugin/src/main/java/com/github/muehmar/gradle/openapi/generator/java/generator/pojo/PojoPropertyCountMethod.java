package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties.additionalPropertiesName;
import static io.github.muehmar.codegenerator.Generator.constant;
import static io.github.muehmar.codegenerator.java.JavaModifier.PUBLIC;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.ValidationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.jackson.JacksonAnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaObjectPojo;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.JavaDocGenerator;
import io.github.muehmar.codegenerator.java.MethodGen;
import io.github.muehmar.codegenerator.java.MethodGenBuilder;
import lombok.Value;

public class PojoPropertyCountMethod {
  private PojoPropertyCountMethod() {}

  private static final String JAVA_DOC = "Returns the number of present properties of this object.";

  public static Generator<JavaObjectPojo, PojoSettings> pojoPropertyCountMethoGenerator() {
    final MethodGen<JavaObjectPojo, PojoSettings> method =
        MethodGenBuilder.<JavaObjectPojo, PojoSettings>create()
            .modifiers(PUBLIC)
            .noGenericTypes()
            .returnType("int")
            .methodName("getPropertyCount")
            .noArguments()
            .content(propertyCountMethodContent())
            .build();
    return JavaDocGenerator.<JavaObjectPojo, PojoSettings>ofJavaDocString(JAVA_DOC)
        .append(ValidationGenerator.minAnnotationForPropertyCount(), JavaObjectPojo::getConstraints)
        .append(ValidationGenerator.maxAnnotationForPropertyCount(), JavaObjectPojo::getConstraints)
        .append(JacksonAnnotationGenerator.jsonIgnore())
        .append(method);
  }

  private static Generator<JavaObjectPojo, PojoSettings> propertyCountMethodContent() {
    final Generator<PojoAndMember, PojoSettings> propertyCount =
        requiredNotNullablePropertyCount()
            .append(requiredNullablePropertyCount())
            .append(optionalNotNullablePropertyCount())
            .append(optionalNullablePropertyCount());
    return Generator.<JavaObjectPojo, PojoSettings>emptyGen()
        .append(w -> w.println("return"))
        .appendList(propertyCount.indent(1), PojoAndMember::fromPojo)
        .append(additionalPropertiesCount().indent(1));
  }

  private static Generator<PojoAndMember, PojoSettings> requiredNotNullablePropertyCount() {
    return Generator.<PojoAndMember, PojoSettings>emptyGen()
        .append((pam, s, w) -> w.println("(%s != null ? 1 : 0) +", pam.getMember().getName()))
        .filter(pam -> pam.getMember().isRequiredAndNotNullable());
  }

  private static Generator<PojoAndMember, PojoSettings> requiredNullablePropertyCount() {
    return Generator.<PojoAndMember, PojoSettings>emptyGen()
        .append((pam, s, w) -> w.println("(%s ? 1 : 0) +", pam.getMember().getIsPresentFlagName()))
        .filter(pam -> pam.getMember().isRequiredAndNullable());
  }

  private static Generator<PojoAndMember, PojoSettings> optionalNotNullablePropertyCount() {
    return Generator.<PojoAndMember, PojoSettings>emptyGen()
        .append((pam, s, w) -> w.println("(%s != null ? 1 : 0) +", pam.getMember().getName()))
        .filter(pam -> pam.getMember().isOptionalAndNotNullable());
  }

  private static Generator<PojoAndMember, PojoSettings> optionalNullablePropertyCount() {
    return Generator.<PojoAndMember, PojoSettings>emptyGen()
        .append(
            (pam, s, w) ->
                w.println(
                    "((%s || %s != null) ? 1 : 0) +",
                    pam.getMember().getIsNullFlagName(), pam.getMember().getName()))
        .filter(pam -> pam.getMember().isOptionalAndNullable());
  }

  private static Generator<JavaObjectPojo, PojoSettings> additionalPropertiesCount() {
    return Generator.<JavaObjectPojo, PojoSettings>emptyGen()
        .append(constant("%s.size();", additionalPropertiesName()));
  }

  @Value
  private static class PojoAndMember {
    JavaObjectPojo pojo;
    JavaPojoMember member;

    private static PList<PojoAndMember> fromPojo(JavaObjectPojo pojo) {
      return pojo.getAllMembers().map(member -> new PojoAndMember(pojo, member));
    }
  }
}
