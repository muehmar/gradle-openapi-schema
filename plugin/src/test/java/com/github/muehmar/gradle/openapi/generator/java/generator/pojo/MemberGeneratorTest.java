package com.github.muehmar.gradle.openapi.generator.java.generator.pojo;

import static com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberGenerator.memberGenerator;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import com.github.muehmar.gradle.openapi.generator.java.JacksonRefs;
import com.github.muehmar.gradle.openapi.generator.java.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.MemberGenerator.MemberContent;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;

@SnapshotTest
class MemberGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("allNecessityAndNullabilityVariants")
  void generate_when_samplePojo_then_correctOutputAndRef() {
    final Generator<MemberContent, PojoSettings> gen = memberGenerator();

    final Writer writer =
        gen.generate(
            JavaPojos.allNecessityAndNullabilityVariants().getMemberContent(),
            defaultTestSettings(),
            javaWriter());

    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_MAP::equals));

    final String output = writer.asString();
    expect.toMatchSnapshot(output);
  }

  @Test
  @SnapshotName("illegalIdentifierPojo")
  void generate_when_illegalIdentifierPojo_then_matchSnapshot() {
    final Generator<MemberContent, PojoSettings> gen = memberGenerator();

    final Writer writer =
        gen.generate(
            JavaPojos.illegalIdentifierPojo().getMemberContent(),
            defaultTestSettings(),
            javaWriter());

    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_MAP::equals));

    final String output = writer.asString();
    expect.toMatchSnapshot(output);
  }

  @Test
  void generate_when_arrayPojo_then_correctOutputAndRef() {
    final Generator<MemberContent, PojoSettings> gen = memberGenerator();

    final Writer writer =
        gen.generate(JavaPojos.arrayPojo().getMemberContent(), defaultTestSettings(), javaWriter());

    assertTrue(writer.getRefs().exists(JacksonRefs.JSON_VALUE::equals));

    final String output = writer.asString();
    assertEquals("@JsonValue\n" + "private final List<Double> value;", output);
  }
}
