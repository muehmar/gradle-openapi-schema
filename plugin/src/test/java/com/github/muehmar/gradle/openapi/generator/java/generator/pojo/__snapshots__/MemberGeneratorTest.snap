allNecessityAndNullabilityVariants=[
private final String requiredStringVal;
private final String requiredNullableStringVal;
private final boolean isRequiredNullableStringValPresent;
private final String optionalStringVal;
private final boolean isOptionalStringValNotNull;
private final String optionalNullableStringVal;
private final boolean isOptionalNullableStringValNull;
private final Map<String, Object> additionalProperties;
]


illegalIdentifierPojo=[
private final String switch_;
private final boolean isSwitchNull;
private final String point_;
private final boolean isPoint_Null;
private final Map<String, Object> additionalProperties;
]